homeassistant/components/dlna_dmr/data.py

```diff
@@ -0,0 +1,94 @@
+"""Data used by this integration."""
+from __future__ import annotations
+
+import asyncio
+from typing import NamedTuple, cast
+
+from async_upnp_client import UpnpEventHandler, UpnpFactory, UpnpRequester
+from async_upnp_client.aiohttp import AiohttpNotifyServer, AiohttpSessionRequester
+
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant
+from homeassistant.helpers import aiohttp_client
+
+from .const import DOMAIN, LOGGER
+
+
+class EventListenAddr(NamedTuple):
+    """Unique identifier for an event listener."""
+
+    port: int  # Listening port, 0 means use an ephemeral port
+    callback_url: str | None
+
+
+class DlnaDmrData:
+    """Storage class for domain global data."""
+
+    lock: asyncio.Lock
+    requester: UpnpRequester
+    upnp_factory: UpnpFactory
+    event_notifiers: dict[EventListenAddr, AiohttpNotifyServer]
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize global data."""
+        self.lock = asyncio.Lock()
+        session = aiohttp_client.async_get_clientsession(hass, verify_ssl=False)
+        self.requester = AiohttpSessionRequester(session, with_sleep=False)
+        self.upnp_factory = UpnpFactory(self.requester, non_strict=True)
+        self.event_notifiers = {}
+
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.async_shutdown)
```

Comment:
I'm not sure how to achieve that, since the data is shared by all `ConfigEntry`s (it's the integration's global data in `hass.data`). Is there an integration-wide `async_shutdown` (complement to `async_setup`) that could be used?

Reply:
When the async entry unload is called check to see how many config entries are still loaded with the same domain, and if it's the last one, call your unloader

Reply:
Thanks for the tip. I've gone with a slightly different resource management method, releasing the event_notifier when the device is disconnected, and cancelling the shutdown listener when there's no more event_notifiers.